presto-hive/src/main/java/com/facebook/presto/hive/metastore/ThriftHiveMetastore.java

```diff
@@ -769,25 +738,103 @@ public void revokeTablePrivileges(String databaseName, String tableName, String
         }
     }
 
+    // GRANT and REVOKE prioritize roles over user names, given a string which is both, a role name and a user.
+    private static HivePrincipal toRolesPrioritizedHivePrincipal(HiveMetastoreClient metastoreClient, String grantee)
+            throws TException
+    {
+        String principalName = grantee.toLowerCase(Locale.US);
+
+        if (grantee.equalsIgnoreCase(PUBLIC_ROLE_NAME) || metastoreClient.getRoleNames().contains(principalName)) {
+            return new HivePrincipal(principalName, ROLE); // Hive metastore API requires role names to be in lowercase.
+        }
+        else {
+            return new HivePrincipal(grantee, USER);
+        }
+    }
+
+    private PrivilegeBag buildPrivilegeBag(
+            String databaseName,
+            String tableName,
+            HivePrincipal hivePrincipal,
+            Set<PrivilegeGrantInfo> privilegeGrantInfos)
+            throws TException
+    {
+        ImmutableList.Builder<HiveObjectPrivilege> privilegeBagBuilder = ImmutableList.builder();
+        for (PrivilegeGrantInfo privilegeGrantInfo : privilegeGrantInfos) {
+            privilegeBagBuilder.add(
+                    new HiveObjectPrivilege(
+                            new HiveObjectRef(TABLE, databaseName, tableName, null, null),
+                            hivePrincipal.getPrincipalName(),
+                            hivePrincipal.getPrincipalType(),
+                            privilegeGrantInfo));
+        }
+        return new PrivilegeBag(privilegeBagBuilder.build());
+    }
+
     private boolean containsAllPrivilege(Set<PrivilegeGrantInfo> requestedPrivileges)
     {
         return requestedPrivileges.stream()
                 .anyMatch(privilege -> privilege.getPrivilege().equalsIgnoreCase("all"));
     }
 
-    private Set<HivePrivilegeInfo> getPrivileges(String user, HiveObjectRef objectReference)
+    private Set<HivePrivilegeInfo> getTablePrivileges(HivePrincipal hivePrincipal, String databaseName, String tableName)
```

Comment:
Yes, it is. This one is a private method where as the other is a public method. I thought about renaming the public method to `getUserTablePrivileges` since it only retrieves privileges of a user. But a better solution might be to create and pass a SPI "Principal" as an argument that can represent both, a user or a role. I can do it in a follow-up PR since this involves more changes.